Refactor `TransferBalancePage` into a functional component

francoisl opened the pull request:
### Details
This refactors `TransferBalancePage` into a functional component.

### Fixed Issues
$ https://github.com/Expensify/App/issues/16288

### Tests

Prerequisites: have two accounts set up with bank accounts for money transfers.

1. From account `A`, send money to account `B`
2. Sign in to account B, and go to Settings > Payment > Transfer Balance
3. Below "Which account?", click on the selected account
4. In the view to select an account, make sure the `default` account is pre-selected
5. Select any other account
6. Make sure you are redirected back to the Transfer Balance page, and the account you just chose is selected


- [x] Verify that no errors appear in the JS console

### Offline tests
N/A, being online is required to navigate to the Transfer Balance page

### QA Steps

Prerequisites: have two accounts set up with bank accounts for money transfers.

1. From account `A`, send money to account `B`
2. Sign in to account B, and go to Settings > Payment > Transfer Balance
3. Below "Which account?", click on the selected account
4. In the view to select an account, make sure the `default` account is pre-selected
5. Select any other account
6. Make sure you are redirected back to the Transfer Balance page, and the account you just chose is selected



- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>


https://github.com/Expensify/App/assets/2229301/cfc93c26-fe7d-49fe-ae5a-4013554e513b


</details>

<details>
<summary>Mobile Web - Chrome</summary>


https://github.com/Expensify/App/assets/2229301/16c136e3-918a-48d1-b08b-6e527aa2c632



</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/2229301/a0b9b157-bf34-4092-9f15-64fdfc7e4a0a



</details>

<details>
<summary>Desktop</summary>


https://github.com/Expensify/App/assets/2229301/58729b7b-0886-4db1-82b5-ed7004b8cb7e


</details>

<details>
<summary>iOS</summary>


https://github.com/Expensify/App/assets/2229301/e7a5fd34-08c0-42b9-9d32-7011dfa37b35


</details>

<details>
<summary>Android</summary>


https://github.com/Expensify/App/assets/2229301/f14fe12a-045a-4223-8438-5a9a91512aae



</details>

francoisl committed:
Refactor `TransferBalancePage` into a functional component

francoisl committed:
Add missing `displayName` property

francoisl committed:
Run prettier

francoisl commented on src/pages/settings/Payments/TransferBalancePage.js:
```diff
@@ -65,65 +65,64 @@ const defaultProps = {
      */
-    getSelectedPaymentMethodAccount() {
-        const paymentMethods = PaymentUtils.formatPaymentMethods(this.props.bankAccountList, this.props.cardList);
+    const getSelectedPaymentMethodAccount = useCallback(() => {
```
👍 ah yeah, thanks for the reminder. To clarify though, you meant this for the whole component and not just this `useCallback`, correct?

marcaaron replied:
Maybe this comment is in the wrong spot? You certainly can declare a function like this if you are calling a function that returns a function.

francoisl committed:
Declare component with the `function` keyword instead of using an arrow function

marcaaron commented on src/pages/settings/Payments/TransferBalancePage.js:
```diff
@@ -65,65 +65,64 @@ const defaultProps = {
      */
-    getSelectedPaymentMethodAccount() {
-        const paymentMethods = PaymentUtils.formatPaymentMethods(this.props.bankAccountList, this.props.cardList);
+    const getSelectedPaymentMethodAccount = useCallback(() => {
```
I _think_ we can probably skip `useCallback()` here according to the latest debate on it 😄

Just inline the function and we will only worry about the optimizations if there's a good reason for it. Or if there's some good reason for it we should explain in a comment what it is. If there is, it wasn't obvious to me.

marcaaron commented on src/pages/settings/Payments/TransferBalancePage.js:
```diff
@@ -65,65 +65,64 @@ const defaultProps = {
+        }
+
+        PaymentMethods.saveWalletTransferAccountTypeAndID(selectedAccount.accountType, selectedAccount.methodID);
+    }, [getSelectedPaymentMethodAccount]);
```
This is... interesting... I am sort of wondering why we are running that logic in the constructor before. Because it would only ever run once for the life of the instantiated component and now it will run again whenever the reference changes (which would happen if any props change) 🤔

Not sure how to advise as I'm not sure why we used the constructor in the first place. But my first instinct would be to move this into a `useEffect()` with an empty array as a dependency since that is the closest thing to running logic in the constructor (but isn't exactly the same). 

I haven't run into this much. It seems uncommon to run logic in the constructor unless you are initializing state.

francoisl replied:
Yeah so for context, at first I was trying to use `[]` for the dependencies here to only run the effect once, but I would get the `exhaustive-deps` ESLint warning. Same thing if I didn't `useCallback()` for `getSelectedPaymentMethodAccount`.

I'll go ahead and try your suggestion though, thanks.

marcaaron commented on src/pages/settings/Payments/TransferBalancePage.js:
```diff
@@ -65,65 +65,64 @@ const defaultProps = {
      * @param {String} filterPaymentMethodType
      */
-    navigateToChooseTransferAccount(filterPaymentMethodType) {
+    const navigateToChooseTransferAccount = (filterPaymentMethodType) => {
```
```suggestion
    function navigateToChooseTransferAccount (filterPaymentMethodType) {
```

francoisl committed:
Pull main, fix conflicts

francoisl committed:
PR comments

- Consolidate into a single `useEffect` with empty array dependency, to only run once on initial render
- Change `useCallback` function into a regular function

francoisl commented:
Updated and retested!

marcaaron commented:
@0xmiroslav can you check this one out and give me a shout when it looks ready? 🙇

francoisl commented:
@0xmiroslav did you get a chance to review the PR again?

francoisl commented:
@0xmiroslav bump please

francoisl commented on src/pages/settings/Payments/TransferBalancePage.js:
```diff
@@ -135,120 +120,132 @@ class TransferBalancePage extends React.Component {
-        const isTransferable = transferAmount > 0;
-        const isButtonDisabled = !isTransferable || !selectedAccount;
-        const errorMessage = !_.isEmpty(this.props.walletTransfer.errors) ? _.chain(this.props.walletTransfer.errors).values().first().value() : '';
+        const selectedAccount = getSelectedPaymentMethodAccount();
+        if (!selectedAccount) {
+            return;
+        }
```
Yeah I think this was a bug, if we return early here and you have a non-default account selected, then we can't properly pre-select the account in the `ChooseTransferAccountPage`.

Here's what it looks like if we add the `|| !selectedAccount.isDefault` (also happening on `main`):

https://github.com/Expensify/App/assets/2229301/a37b5031-22bd-4271-8efc-22d8bef3925f

francoisl commented:
Good question. Looking at previous PRs that modified this component, they were mostly tested internally. I asked in Slack [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1686960280953209) for now.

marcaaron commented:
@francoisl I'm gonna jump in and take over the review on this one as @0xmiroslav may not be able to test. 

I think to get to that page you basically would need to have:

- gold wallet
- paid via Expensify for an IOU
- bank account to transfer to
- etc

So, it seems quite hard for a C+ to test actually. I am _guessing_ that's the explanation for the delay here.

marcaaron commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

marcaaron commented:
This PR was a bit broken for me in the current state - but was fixed after merging in `main`. After sending money the app was crashing. I think it's unrelated to these changes though.

One thing I have noticed (also unrelated) - the button here does not go anywhere... though this feature is still behind a beta IIRC so we are good. Will continue testing on all the platforms now!

![2023-06-27_09-10-51](https://github.com/Expensify/App/assets/32969087/3286194b-b87f-406c-a360-1718573afbd5)

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

marcaaron commented:
Whoops was I not supposed to merge this? I heard something about a "merge freeze" - hopefully it is OK.

francoisl commented:
Haha I forgot about that too. I think it's ok that we merged though.

kbecciv commented:
Hello @francoisl @marcaaron QA team have trouble to see the Transfer balance option under Payment. Can you please help us? Is this internal testing? Thank you

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5416191215) to production by https://github.com/thienlnam in version: 1.3.34-1 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
